### PR TITLE
feat: Add tooltip for wrong decimal format in Text Entry

### PR DESCRIPTION
**Pull request by taopkorczak**
Related to: https://oat-sa.atlassian.net/browse/TR-6116

When the Text Entry input field is blurred:

A red border to the input is displayed.

A tooltip is displayed above the input, mentioning the accepted number format for the input (adapted for configured decimalSeparator and thousandsSeparator), for example Invalid value, use . (dot) for decimal point and , (comma) for thousands separator.
Zrzut ekranu 2024-05-22 o 08 02 02
Zrzut ekranu 2024-05-22 o 08 02 36
The value is still truncated if submitted as an invalid value.

### CODE REVIEW

File: src/qtiCommonRenderer/renderers/interactions/TextEntryInteraction.js [hunk 1]
```diff
@@ -72,6 +72,29 @@ function showTooltip($input, theme, message) {
...
+ *
+ * @param {jQuery} $input
+ */
+function validateDecimalInput ($input) {
+    const value = $input.val();
+    const regex = /^$|^-?\d{1,3}(,\d{3})*(\.\d+)?$|^-?\d+(\.\d+)?$|^-?\d*\.$|^-?\.\d+$/;
+    if (!regex.test(value)) {
+        $input.addClass('invalid');
+        $input.addClass('error');
+        showTooltip($input, 'error', __('Invalid value, use . (dot) for decimal point and , (comma) for thousands separator.'));
```
Comment (olga-kulish) [inline src/qtiCommonRenderer/renderers/interactions/TextEntryInteraction.js:90]: Just a note: the ticket says: `The value is still truncated if submitted as an invalid value.`3,14` is submitted as `3`. But I think we should submit `null`, as we do in Solar? Anyway, this is out of scope in this ticket.

**Review by olga-kulish — COMMENTED** [review]
Cool idea to check that it's actually a 'thousand' (3 digits)

File: src/qtiCommonRenderer/renderers/interactions/TextEntryInteraction.js (hunk 1 above)
Comment (taopkorczak) [reply src/qtiCommonRenderer/renderers/interactions/TextEntryInteraction.js:91]: In task they said: 'The value is still truncated if submitted as an invalid value.', thats why I didn't change this logic.

File: src/qtiCommonRenderer/renderers/interactions/TextEntryInteraction.js [hunk 2]
```diff
@@ -72,6 +72,29 @@ function showTooltip($input, theme, message) {
...
+ *
+ * This function ensures that the input value is either empty or follows
+ * the rules for decimal numbers. It allows numbers with optional
+ * thousands separators (commas) and a mandatory decimal point (dot).
+ *
+ * @param {jQuery} $input
+ */
+function validateDecimalInput ($input) {
+    const value = $input.val();
+    const regex = /^$|^-?\d{1,3}(,\d{3})*(\.\d+)?$|^-?\d+(\.\d+)?$|^-?\d*\.$|^-?\.\d+$/;
```
[the remark this answers is not in the document]
Comment (taopkorczak) [reply src/qtiCommonRenderer/renderers/interactions/TextEntryInteraction.js]: Thank you. I fixed it. Now I'm getting `thousandsSeparator` and `decimalSeparator` from configuration. Also I noticed one more case, that thousandsSeparator can be empty, decimalSeparator will be return '.' if will be not defined. So I'm checking if thousandsSeparator is empty, than I'm created regex without verification of thousands and display error `Invalid value, use . (dot) for decimal point.` if is defined than making verification and display error `Invalid value, use . (dot) for decimal point and , (comma) for thousands separator.`

File: src/qtiCommonRenderer/renderers/interactions/TextEntryInteraction.js [hunk 3]
```diff
@@ -72,6 +72,43 @@ function showTooltip($input, theme, message) {
...
+        ? `^$|^-?\\d{1,3}(${escapedThousandsSeparator}\\d{3})*(${escapedDecimalSeparator}\\d+)?$|^-?\\d+(${escapedDecimalSeparator}\\d+)?$|^-?\\d*${escapedDecimalSeparator}$|^-?${escapedDecimalSeparator}\\d+$`
+        : `^$|^-?\\d+(${escapedDecimalSeparator}\\d+)?$|^-?\\d*${escapedDecimalSeparator}$|^-?${escapedDecimalSeparator}\\d+$`;
+
+    const regex = new RegExp(regexPattern);
+
+    if (!regex.test(value)) {
+        $input.addClass('invalid');
+        $input.addClass('error');
+        const decimalError = thousandsSeparator
+            ? __('Invalid value, use . (dot) for decimal point and , (comma) for thousands separator.')
```
Comment (olga-kulish) [inline src/qtiCommonRenderer/renderers/interactions/TextEntryInteraction.js:102]: Separator in the message is still hardcoded.

File: src/qtiCommonRenderer/renderers/interactions/TextEntryInteraction.js [hunk 4]
```diff
@@ -72,6 +72,43 @@ function showTooltip($input, theme, message) {
...
+        ? `^$|^-?\\d{1,3}(${escapedThousandsSeparator}\\d{3})*(${escapedDecimalSeparator}\\d+)?$|^-?\\d+(${escapedDecimalSeparator}\\d+)?$|^-?\\d*${escapedDecimalSeparator}$|^-?${escapedDecimalSeparator}\\d+$`
+        : `^$|^-?\\d+(${escapedDecimalSeparator}\\d+)?$|^-?\\d*${escapedDecimalSeparator}$|^-?${escapedDecimalSeparator}\\d+$`;
+
+    const regex = new RegExp(regexPattern);
+
+    if (!regex.test(value)) {
+        $input.addClass('invalid');
+        $input.addClass('error');
+        const decimalError = thousandsSeparator
+            ? __('Invalid value, use %s (dot) for decimal point and %s (comma) for thousands separator.', decimalSeparator, thousandsSeparator)
```
Comment (olga-kulish) [inline src/qtiCommonRenderer/renderers/interactions/TextEntryInteraction.js:102]: Still not quite right. If separators were set from `ru-RU`, text would be: "_Invalid value, use , (dot) for decimal point and . (comma) for thousands separator._". Symbol-name should correspond to the symbol. In case of unknown symbol, write no symbol-name at all.

File: src/qtiCommonRenderer/renderers/interactions/TextEntryInteraction.js [hunk 5]
```diff
@@ -72,6 +72,49 @@ function showTooltip($input, theme, message) {
...
+ * This function ensures that the input value is either empty or follows
+ * the rules for decimal numbers. It allows numbers with optional
+ * thousands separators (commas) and a mandatory decimal point (dot).
+ *
+ * @param {jQuery} $input
+ */
+function validateDecimalInput ($input) {
+    const separatorName = {
+        '.': __('(dot)'),
+        ',': __('(comma)')
```
Comment (olga-kulish) [inline src/qtiCommonRenderer/renderers/interactions/TextEntryInteraction.js:87]: Would be nice to add `' ': __('(space)')`, because in theory it could be set as thousands separator, and looks weird in that case.

**Review by olga-kulish — COMMENTED** [review]
Works nicely.

But unit tests are failing, please fix them. 
Would be nice to add a test for these changes too, unless it's difficult.

PS: Don't forget to `npm i/npm run build` before `npm run test`, otherwise code changes won't be applied.

**Review by olga-kulish — APPROVED** [review]
